### PR TITLE
Budgeting

### PR DESCRIPTION
**Pull request by aumayr**

### CODE REVIEW

**Comment by yagebu** [comment]
Did you force-push this? Seems my last three commits to master were removed...

**Comment by aumayr** [comment]
Yes, but I did a git-pull first, so they should be there… Sorry!

**Comment by aumayr** [comment]
Damn, your three commits are really gone… any way to recover them? 😞

File: fava/api/__init__.py [hunk 1]
```diff
@@ -295,7 +302,33 @@ def interval_balances(self, interval, account_name, accumulate=False):
...
+
+    def add_budgets(self, zipped_interval_balances, interval_tuples, accumulate):  # noqa
+        """Add budgets data to zipped (recursive) interval balances."""
+        if not zipped_interval_balances:
+            return
+
+        interval_budgets = [self.budgets.budget(
+                zipped_interval_balances['account'],
+                interval_tuples[0][0] if accumulate else begin_date,
+                end_date - rdelta(days=1)
```
Comment (yagebu) [inline fava/api/__init__.py]: Does this mean that the dates that the budget function accepts are inclusive? If so it should be changed to exclusive as that's the way the current code and beancount does it.

Also: a simple `timedelta(1)` can be used here.

File: fava/api/__init__.py [hunk 2]
```diff
@@ -23,6 +23,9 @@
 import operator
 import os
 
+from dateutil.relativedelta import relativedelta as rdelta
+from decimal import Decimal
```
Comment (yagebu) [inline fava/api/__init__.py]: Decimal should be imported from beancount I guess.

**Comment by yagebu** [comment]
> Damn, your three commits are really gone… any way to recover them? 😞

I still have them, so I'll rebase them and push them again.

File: fava/util/date.py [hunk 3]
```diff
@@ -169,3 +170,36 @@ def parse_date(string):
...
+def number_of_days_in_period(period, date_):
+    """Returns the days in the specified period and date_."""
+
+    if period == 'daily':
+        return 1
+    if period == 'weekly':
+        return 7
+    if period == 'monthly':
+        date_ = datetime.datetime(date_.year, date_.month, 1)
+        return ((date_ + relativedelta(months=1)) - date_).days
```
Comment (yagebu) [inline fava/util/date.py]: Use `get_next_interval`

File: fava/util/date.py [hunk 4]
```diff
@@ -169,3 +170,36 @@ def parse_date(string):
...
+        quarter = (date_.month - 1) / 3 + 1
+        if quarter == 1:
+            date_ = datetime.datetime(date_.year, 1, 1)
+        if quarter == 2:
+            date_ = datetime.datetime(date_.year, 4, 1)
+        if quarter == 3:
+            date_ = datetime.datetime(date_.year, 7, 1)
+        if quarter == 4:
+            date_ = datetime.datetime(date_.year, 10, 1)
+        return ((date_ + relativedelta(months=3)) - date_).days
```
Comment (yagebu) [inline fava/util/date.py]: Use `get_next_interval`

File: fava/util/date.py [hunk 5]
```diff
@@ -169,3 +170,36 @@ def parse_date(string):
...
+    if period == 'yearly':
+        date_ = datetime.datetime(date_.year, 1, 1)
+        return ((date_ + relativedelta(years=1)) - date_).days
```
Comment (yagebu) [inline fava/util/date.py]: Use `get_next_interval`

**Comment by aumayr** [comment]
Boah, good to hear… sorry again! 😞

File: fava/util/date.py [hunk 6]
```diff
@@ -169,3 +170,36 @@ def parse_date(string):
...
+    if period == 'daily':
+        return 1
+    if period == 'weekly':
+        return 7
+    if period == 'monthly':
+        date_ = datetime.datetime(date_.year, date_.month, 1)
+        return ((date_ + relativedelta(months=1)) - date_).days
+    if period == 'quarterly':
+        quarter = (date_.month - 1) / 3 + 1
+        if quarter == 1:
```
Comment (yagebu) [inline fava/util/date.py]: date_ = datetime.datetime(date_year, quarter*3 -2, 1)

File: fava/util/date.py [hunk 7]
```diff
@@ -169,3 +170,36 @@ def parse_date(string):
             return daterange(m.year, m.month)
         else:
             return daterange(today.year + modifier, _parse_month(identifier))
+
+
+def days_in_daterange(start_date, end_date):
+    """Yields a datetime for every day in the specified interval."""
+    for n in range(int((end_date - start_date).days) + 1):
+        yield start_date + relativedelta(days=n)
```
Comment (yagebu) [inline fava/util/date.py]: `relativedelta` -> `timedelta`

**Comment by yagebu** [comment]
> Boah, good to hear… sorry again! 😞

No worries ;)

I've added some comments to the code. Overall this looks good and I'll have to try it some time 👍

**Comment by aumayr** [comment]
Thanks, will work on them right away! 

I plan to add more charts and even a separate "Budgets" page, but first this has to go into master, as it should go out with the next release for people to play around with.

**Pull request merged**